### PR TITLE
Feature/pretraining-auto-truncate

### PR DESCRIPTION
**Pull request by Blyzi**
# Description

This PR adds auto-truncation support for pretraining_datasets.

## Motivation and Context

While investigating an issue in the [dataloader](#3069), I noticed an opportunity to improve the handling of pretraining_datasets.
Although this PR is not a direct fix for that issue, adding auto-truncation was straightforward.

One important thing to notice is that it will become harder for the user to know the `max_steps` value accordingly to its datasets because it's no longer the number of samples.

## How has this been tested?

I tested the changes locally on my setup with multiple GPUs.
The modification is scoped only to pretraining_datasets with or without concatenation, and only when sample packing is disabled.
While I did not perform extensive testing, I validated the tokenisation process.

## Types of changes

- [x] Enhancement (non-breaking change that adds functionality)

## Social Handles

[Blyzi](https://github.com/Blyzi)

## Summary by CodeRabbit

- Refactor
  - Pretraining data encoding now produces fixed-size token chunks without truncating inputs or adding EOS/PAD tokens; optional concatenation merges samples before chunking.
  - Padding and attention masks are applied consistently; labels ignore padded positions.

- Documentation
  - Removed callout about automatic splitting of long completion samples and updated related examples in dataset format docs.

### CODE REVIEW

File: src/axolotl/utils/data/pretraining.py [hunk 1]
```diff
@@ -24,157 +24,81 @@ def encode_pretraining(
...
-                        tokenizer.pad_token_id,
-                        dtype=torch.long,
-                    ),
-                ),
-                dim=0,
+    inputs_ids, target_ids, attention_mask = [], [], []
+
+    # Concatenate if specified all input_ids and attention masks into one tensor when concatenate is True
+    if concatenate:
+        full_inputs["input_ids"] = [torch.cat(full_inputs["input_ids"], dim=0)]
```
Comment (djsaunde) [inline src/axolotl/utils/data/pretraining.py:57]: Dumb question, not too familiar with this codepath: is this not the same as sample packing?

Comment (Blyzi) [reply src/axolotl/utils/data/pretraining.py:57]: "Dumb question", don't worry IA is moving way too fast 😄 

The main objective of sample packing is to avoid attention noise between samples while filling the sequence length. It has been proven to improve model performance while maintaining computational speed. Historically, during pre-training, we didn't care about noise between samples, so we simply concatenated everything.

Comment (winglian) [reply src/axolotl/utils/data/pretraining.py:57]: in pretraining, this can already be accomplished with `sample_packing: true` and `pretrain_multipack_attn: false`

Comment (winglian) [reply src/axolotl/utils/data/pretraining.py:57]: it's an underdocumented feature, but should be equivalent to no sample packing with concatenate, but with more efficient packing

Comment (Blyzi) [reply src/axolotl/utils/data/pretraining.py:57]: Do you guys need me to do anything before merging it ?
